mlflow/utils/autologging_utils/client.py

```diff
@@ -0,0 +1,325 @@
+            tags, max_key_length=MAX_ENTITY_KEY_LENGTH, max_value_length=MAX_TAG_VAL_LENGTH
+        )
+        tags_arr = [RunTag(key, str(value)) for key, value in tags.items()]
+        self._get_pending_operations(run_id).enqueue(tags=tags_arr)
+
+    def flush(self, synchronous=True):
+        """
+        Flushes all queued run operations, resulting in the creation or mutation of runs
+        and run data.
+
+        :param synchronous: If `True`, run operations are performed synchronously, and a
+                            `RunOperations` result object is only returned once all operations
+                            are complete. If `False`, run operations are performed asynchronously,
+                            and an `RunOperations` object is returned that represents the ongoing
+                            run operations.
+        :return: A `RunOperations` instance representing the flushed operations. These operations
+                 are already complete if `synchronous` is `True`. If `synchronous` is `False`, these
+                 operations may still be inflight. Operation completion can be synchronously waited
+                 on via `RunOperations.await_completion()`.
+        """
+        logging_futures = []
+        for pending_operations in self._pending_ops_by_run_id.values():
+            future = self._thread_pool.submit(
+                self._flush_pending_operations, pending_operations=pending_operations,
+            )
+            logging_futures.append(future)
+        self._pending_ops_by_run_id = {}
+
+        logging_operations = RunOperations(logging_futures)
+        if synchronous:
+            logging_operations.await_completion()
+        return logging_operations
+
+    def _get_pending_operations(self, run_id):
+        """
+        :return: A `_PendingRunOperations` containing all pending operations for the specified
+        """
+        if run_id not in self._pending_ops_by_run_id:
+            self._pending_ops_by_run_id[run_id] = _PendingRunOperations(run_id=run_id)
+        return self._pending_ops_by_run_id[run_id]
+
+    def _try_operation(self, fn, *args, **kwargs):
+        """
+        Attempt to evaluate the specified function, `fn`, on the specified `*args` and `**kwargs`,
+        returning either the result of the function evaluation (if evaluation was successful) or
+        the exception raised by the function evaluation (if evaluation was unsuccesful).
+        """
+        try:
+            return fn(*args, **kwargs)
+        except Exception as e:
+            return e
+
+    def _flush_pending_operations(self, pending_operations):
+        """
+        Synchronously and sequentially flushes the specified list of pending run operations.
+
+        NB: Operations are not parallelized on a per-run basis because MLflow's File Store, which
+        is frequently used for local ML development, does not support threadsafe metadata logging
+        within a given run.
```

Comment:
This is fairly obnoxious. We could consider special-casing non-file-store tracking URIs. We've had discussions in the past about deprecating the file store, but we haven't yet taken that step.

For now, I'd opt for synchronous / sequential execution of logging within a run. This also reduces the likelihood of saturating / overloading the Databricks MLflow Tracking server (and other third-party Tracking implementations).

Reply:
This is probably fine for now. For file store we could potentially introduce chunked files to support concurrent write.